### PR TITLE
First exchange of Vircadia protocol packets with the Vircadia Web SDK

### PR DESCRIPTION
**Pull request by ctrlaltdavid**
This PR in conjunction with Vircadia-Web-SDK PR https://github.com/vircadia/vircadia-web-sdk/pull/9 implements the first exchange of Vircadia protocol packets with the domain server.

Merge sequence:
- [x] Vircadia PR: https://github.com/vircadia/vircadia/pull/1238
- [x] Update this PR with the latest `webapp` branch.
- [x] Vircadia-Web-SDK PR: https://github.com/vircadia/vircadia-web-sdk/pull/9
- [x] Update Vircadia-Web repo to use the latest Vircadia-Web-SDK.
- [x] Update this PR to use the latest Vircadia-Web master branch.
- [ ] CR, QA.
- [ ] Merge this PR.

### CODE REVIEW

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-1aa1f87-_2a9ab98e.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1aa1f87-_2a9ab98e.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1aa1f87-_2a9ab98e.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1aa1f87-.tar_2a9ab98e.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1aa1f87-.tar_2a9ab98e.Z

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1aa1f87-_2a9ab98e.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-544c6f9-.tar_4d375c02.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-544c6f9-.tar_4d375c02.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-544c6f9-_4d375c02.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-544c6f9-_4d375c02.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-544c6f9-_4d375c02.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-544c6f9-_4d375c02.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-.tar_b7fdd5bb.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-.tar_b7fdd5bb.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-_b7fdd5bb.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-040906e-_b7fdd5bb.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-_b7fdd5bb.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-_b7fdd5bb.exe

File: libraries/networking/src/AddressManager.h [hunk 1]
```diff
@@ -252,8 +252,8 @@ public slots:
      *     <code>"127.0.0.1"</code> or <code>"localhost"</code>), a <code>file:///</code> address, a domain name, a named path 
      *     on a domain (starts with <code>"/"</code>), a position or position and orientation, or a user (starts with 
      *     <code>"@"</code>).
-     * @param {boolean} [fromSuggestions=false] - Set to <code>true</code> if the address is obtained from the "Goto" dialog.
-     *     Helps ensure that user's location history is correctly maintained.
+     * @param {boolean} [fromSuggestions=false] - Set to <code>true</code> if the address is obtained from the "Explore" app.
+     *     Helps ensure that the user's location history is correctly maintained.
```
Comment (daleglass) [inline libraries/networking/src/AddressManager.h:256]: This change probably doesn't belong in this PR

File: libraries/networking/src/LimitedNodeList.cpp [hunk 2]
```diff
@@ -204,15 +206,19 @@ void LimitedNodeList::setPermissions(const NodePermissions& newPermissions) {
...
     }
-    if (_nodeSocket.localPort() != socketLocalPort) {
-        _nodeSocket.rebind(socketLocalPort);
-        LIMITED_NODELIST_LOCAL_PORT.set(socketLocalPort);
+    if (_nodeSocket.localPort(socketType) != socketLocalPort) {
+        _nodeSocket.rebind(socketType, socketLocalPort);
+        if (socketType == SocketType::UDP) {
+            LIMITED_NODELIST_LOCAL_PORT.set(socketLocalPort);
+        } else {
+            // WEBRTC TODO: Add WebRTC equivalent?
```
Comment (daleglass) [inline libraries/networking/src/LimitedNodeList.cpp:220]: Perhaps add a log message here in case anything ends up getting here?

File: libraries/networking/src/udt/NetworkSocket.cpp [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    }
+}
+
+
+qint64 NetworkSocket::writeDatagram(const QByteArray& datagram, const HifiSockAddr& sockAddr) {
+    switch (sockAddr.getSocketType()) {
+    case SocketType::UDP:
+        // WEBRTC TODO: The Qt documentation says that the following call shouldn't be used if the UDP socket is connected!!!
+        // https://doc.qt.io/qt-5/qudpsocket.html#writeDatagram
+        return _udpSocket.writeDatagram(datagram, sockAddr.getAddress(), sockAddr.getPort());
```
Comment (daleglass) [inline libraries/networking/src/udt/NetworkSocket.cpp:130]: Why use `writeDatagram` anyway, against the recommendation, then? It seems easy to call `write` here as recommended.

File: libraries/networking/src/udt/NetworkSocket.h [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+#if defined(WEBRTC_DATA_CHANNELS)
+#include "../webrtc/WebRTCSocket.h"
+#endif
+
+/// @addtogroup Networking
+/// @{
+
+
+/// @brief Multiplexes a QUdpSocket and a WebRTCSocket so that they appear as a single QUdpSocket-style socket.
+class NetworkSocket : public QObject {
```
Comment (daleglass) [inline libraries/networking/src/udt/NetworkSocket.h:29]: perhaps this should inherit from QAbstractSocket instead?

File: libraries/networking/src/webrtc/WebRTCSocket.h [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+#include "WebRTCDataChannels.h"
+#include "WebRTCSignalingServer.h"
+
+/// @addtogroup Networking
+/// @{
+
+
+/// @brief Provides a QUdpSocket-style interface for using WebRTCDataChannels.
+class WebRTCSocket : public QObject {
```
Comment (daleglass) [inline libraries/networking/src/webrtc/WebRTCSocket.h:28]: Perhaps it should inherit from QAbstractSocket?

File: tools/doxygen/Doxyfile [hunk 6]
```diff
@@ -509,7 +509,7 @@ EXTRACT_PACKAGE        = NO
 # included in the documentation.
 # The default value is: NO.
 
-EXTRACT_STATIC         = NO
+EXTRACT_STATIC         = YES
```
Comment (daleglass) [inline tools/doxygen/Doxyfile:512]: This probably doesn't belong in this PR

File: tools/doxygen/Doxyfile [hunk 7]
```diff
@@ -794,7 +794,7 @@ CITE_BIB_FILES         =
 # messages are off.
 # The default value is: NO.
 
-QUIET                  = NO
+QUIET                  = YES
```
Comment (daleglass) [inline tools/doxygen/Doxyfile:797]: And this

File: tools/doxygen/README.md [hunk 8]
```diff
@@ -7,8 +7,6 @@
 
 **Doxygen** &ge; 1.9.1 - https://www.doxygen.nl/
 
-Make a `/build/doxygen/` directory.
```
Comment (daleglass) [inline tools/doxygen/README.md:10]: And this

**Review by daleglass — CHANGES_REQUESTED** [review]

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-.tar_b7fdd5bb.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-.tar_b7fdd5bb.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-_b7fdd5bb.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-040906e-_b7fdd5bb.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-_b7fdd5bb.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-040906e-_b7fdd5bb.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1af8c0b-.tar_f02ffe1e.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1af8c0b-.tar_f02ffe1e.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1af8c0b-_f02ffe1e.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-1af8c0b-_f02ffe1e.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1af8c0b-_f02ffe1e.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1af8c0b-_f02ffe1e.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-1af8c0b_f02ffe1e.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Quest-Alpha-PR1259-1af8c0b_f02ffe1e.apk

File: libraries/networking/src/LimitedNodeList.cpp (hunk 2 above)
Comment (ctrlaltdavid) [reply libraries/networking/src/LimitedNodeList.cpp:220]: Done

File: libraries/networking/src/udt/NetworkSocket.cpp (hunk 3 above)
Comment (ctrlaltdavid) [reply libraries/networking/src/udt/NetworkSocket.cpp:133]: This is the way it's currently done in `master` for UDP packets and should be addressed there rather than here.
https://github.com/vircadia/vircadia/issues/1279

File: libraries/networking/src/udt/NetworkSocket.h (hunk 4 above)
Comment (ctrlaltdavid) [reply libraries/networking/src/udt/NetworkSocket.h:29]: `QAbstractSocket` inherits from `QIODevice` and no code from either of these is needed. It's not clear to me whether inheriting from `QAbstractSocket` would be beneficial or just overhead and a potential source of problems.

File: libraries/networking/src/webrtc/WebRTCSocket.h (hunk 5 above)
Comment (ctrlaltdavid) [reply libraries/networking/src/webrtc/WebRTCSocket.h:28]: Ditto.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-0e3627e-_28b46ed8.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-_28b46ed8.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-.tar_28b46ed8.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-.tar_28b46ed8.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-_28b46ed8.sh

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Quest-Alpha-PR1259-0e3627e_28b46ed8.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e_28b46ed8.apk

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-.tar_28b46ed8.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-.tar_28b46ed8.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-_28b46ed8.sh

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-_28b46ed8.dmg

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-0e3627e-_28b46ed8.dmg

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Quest-Alpha-PR1259-0e3627e_28b46ed8.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e_28b46ed8.apk

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-0e3627e-_28b46ed8.exe

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-c8c06b0-_837f2e06.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-.tar_837f2e06.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-.tar_837f2e06.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-_837f2e06.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-_837f2e06.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Quest-Alpha-PR1259-c8c06b0_837f2e06.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0_837f2e06.apk

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-.tar_837f2e06.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-.tar_837f2e06.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-_837f2e06.sh

**build (macOS-latest, client)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Interface-PR1259-c8c06b0-_837f2e06.dmg

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-_837f2e06.dmg

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0-_837f2e06.exe

**build (ubuntu-18.04, android)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-Quest-Alpha-PR1259-c8c06b0_837f2e06.apk
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/vircadia_vircadia/PR1259/Vircadia-PR1259-c8c06b0_837f2e06.apk

**Review by daleglass — APPROVED** [review]

**Review by digisomni — APPROVED** [review]
Worked fine according to the unit tests on master of the Vircadia Web SDK as of 7/15/2021.
